service/matching/taskListManager.go

```diff
@@ -42,20 +42,95 @@ const (
 	done time.Duration = -1
 )
 
+// NOTE: Is this good enough for stress tests?
+const (
+	_maxDispatchDefault = 100000.0
+	_dispatchLimitTTL   = time.Second
+)
+
 type taskListManager interface {
 	Start() error
 	Stop()
 	AddTask(execution *s.WorkflowExecution, taskInfo *persistence.TaskInfo) error
 	GetTaskContext(ctx context.Context) (*taskContext, error)
 	SyncMatchQueryTask(ctx context.Context, queryTask *queryTaskInfo) error
 	CancelPoller(pollerID string)
+	UpdateMaxDispatch(maxDispatchPerSecond *float64)
 	String() string
 }
 
-func newTaskListManager(e *matchingEngineImpl, taskList *taskListID, config *Config) taskListManager {
+type rateLimiter struct {
+	maxDispatchPerSecond *float64
+	tokenBucket          common.TokenBucket
+	ttl                  *time.Timer
+	sync.RWMutex
+	ttlRaw time.Duration
+}
+
+func newRateLimiter(maxDispatchPerSecond *float64, ttl time.Duration) rateLimiter {
+	return rateLimiter{
+		maxDispatchPerSecond: maxDispatchPerSecond,
+		tokenBucket: common.NewTokenBucket(
+			int(*maxDispatchPerSecond), common.NewRealTimeSource(),
+		),
+		ttlRaw: ttl,
+		ttl:    time.NewTimer(ttl),
+	}
+}
+
+func (rl *rateLimiter) UpdateMaxDispatch(maxDispatchPerSecond *float64) {
+	if rl.shouldUpdate(maxDispatchPerSecond) {
+		rl.Lock()
+		defer rl.Unlock()
+		rl.maxDispatchPerSecond = maxDispatchPerSecond
+		rl.tokenBucket = common.NewTokenBucket(int(*maxDispatchPerSecond), common.NewRealTimeSource())
+	}
+}
+
+func (rl *rateLimiter) shouldUpdate(maxDispatchPerSecond *float64) bool {
+	if maxDispatchPerSecond == nil {
+		return false
+	}
+	select {
+	case <-rl.ttl.C:
+		rl.ttl.Reset(rl.ttlRaw)
+		return true
+	default:
+		rl.RLock()
+		defer rl.RUnlock()
+		return *maxDispatchPerSecond < *rl.maxDispatchPerSecond
+	}
+}
+
+func (rl *rateLimiter) TryConsume(count int) (bool, time.Duration) {
+	rl.RLock()
+	defer rl.RUnlock()
+	return rl.tokenBucket.TryConsume(count)
+}
+
+func (rl *rateLimiter) Consume(count int, timeout time.Duration) bool {
+	rl.RLock()
```

Comment:
don't like the fact that we hold the lock while we wait on a caller provided timeout. i.e. if there are 20 go-routines, 19 blocked on Consume() and one blocked on UpdateMaxDispatch(), the latter can starve for a long time (and meanwhile, context deadline could even be exceeded). Here is an alternative -  don't hold the lock within the Consume / TryConsumer methods, instead use atomic.Value to hold the token bucket reference. Whenever you change the token bucket, just atomically set the Value to the new token bucket. That way, callers who hold the old token bucket will stick to the old limit. Newer callers will use the new bucket.

Reply:
Interesting. I would say let's not optimize for that yet and keep it simple, changes are usually rare. Definitely something we can revisit if we hit performance issues.

Reply:
madhu - this is more than just an optimization. Its a bit of correctness as well, in that, we can block the caller for way more than the deadline on a lock contention. With atomic.Value, you will end up with exactly the same lines of code, so IMO, let's just do it right the first time around and forget about this.

Reply:
And fyi - its not about the "value changing" - UpdateMaxDispatch is called regardless for every Poll() request

Reply:
To clarify, it will not affect correctness. Remember that all Poll's will converge on the same throttling value. If one Poll times out, the next one will update it. Also it's not about lines of code, locks are far more common and easier to understand and don't involve type casts. Good read https://texlution.com/post/golang-lock-free-values-with-atomic-value/

Also yes UpdateMaxDispatch is called for every Poll, but write locks are only acquired if the value is changing. Otherwise, only read locks are acquired. Check the shouldUpdate function.

Reply:
I understand that locks are more more common / standard pattern, but atomic.Values aren't that complicated either. In this case, its clearly the right fit IMO. As for correctness, I understand polls will converge, but you will block the caller for more than the caller provided timeout (deadline).  Anyways, will leave it up to you.

> Also yes UpdateMaxDispatch is called for every Poll, but write locks are only acquired if the value is changing
yup, you are right.